base_layer/wallet/src/output_manager_service/service.rs

```diff
@@ -748,6 +759,110 @@ where
         Ok(self.db.get_invalid_outputs().await?)
     }
 
+    pub async fn create_coin_split(
+        &mut self,
+        amount_per_split: MicroTari,
+        split_count: usize,
+        fee_per_gram: MicroTari,
+        lock_height: Option<u64>,
+    ) -> Result<(u64, Transaction, MicroTari, MicroTari), OutputManagerError>
+    {
+        trace!(
+            target: LOG_TARGET,
+            "Select UTXOs and estimate coin split transaction fee."
+        );
+        let total_split_amount = amount_per_split * split_count as u64;
+        let mut estimate_input_count: usize = 1;
+        let mut estimate_output_count = split_count;
+        let mut inputs = Vec::<UnblindedOutput>::new();
+        let mut estimated_fee = MicroTari::default();
+        let mut utxo_total = MicroTari::default();
+        loop {
+            estimated_fee = Fee::calculate(fee_per_gram, 1, estimate_input_count, estimate_output_count);
+            let total_amount = total_split_amount + estimated_fee;
+            inputs = self
+                .select_outputs(total_amount, 0.into(), UTXOSelectionStrategy::MaturityThenSmallest)
+                .await?;
+            // Check if estimated inputs correct
+            let input_count = inputs.len();
+            if input_count != estimate_input_count {
+                estimate_input_count = input_count;
+                continue;
+            }
+            // Check if estimated outputs correct
+            utxo_total = inputs.iter().fold(MicroTari::from(0), |acc, x| acc + x.value);
+            let output_count = if utxo_total >= total_amount {
```

Comment:
utxo_total is guaranteed >= total_amount  here, otherwise it would have returned earlier with an error, so this is really determining whether we must add a change output.

This calc is already done in `select_outputs`, so it feels like we could improve readability AND save CPU cycles by returning a boolean from `select_outputs` as to whether a change output should be added or not.

P.S. I think it should be `if utxo_total > total_amount`...

Reply:
I did not previously do that as the select_outputs method did not use the number of required outputs in its fee calculation. I think your suggestion is better and rather changed the select_outputs method to correctly calculate the fees based on the number of outputs required and then return if a change output is required. That simplified this logic dramatically.

CjS77 are right it should have been ">"